### PR TITLE
Conflicting beforeView forgiveness for Relations

### PR DESCRIPTION
**Pull request by lucaspcram**
### Description:

Since `Relation` views may differ across shards, we need some way to forgive inconsistent `RelationBean` beforeViews when merging relevant `FeatureChange`s. This PR provides that functionality, as well as additional tests.

Still to come: we must implement similar logic for `Node` `inEdge`/`outEdge` sets. This will be added in a future PR.

Additionally, we have implemented beforeView merging for `REMOVE`, since now there are cases (`Relation` and `Node`) where we occasionally must rectify inconsistent beforeViews.

~**DNM Notes:** We may need to rectify the fact that `FeatureChange` `REMOVE`s are no longer a simple merge case, since before views may conflict. For instance, suppose `Relation` 12 is removed by two different `FeatureChanges` which were generated on two different shards. If those shards have different views of `Relation` 12, then we need to merge those beforeViews properly in the `REMOVE` case. Let's not merge until this is rectified.~

### Potential Impact:

Code that previously failed with merge conflicts may now pass.

### Unit Test Approach:

See additional unit tests.

### Test Results:

Tests pass.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java [hunk 1]
```diff
@@ -272,6 +109,92 @@ else if (afterEntityLeft instanceof Relation)
...
+            final RelationBean rightOsmMembers = beforeRelationRight.allKnownOsmMembers().asBean();
+
+            if (!leftMembers.equalsIncludingExplicitlyExcluded(rightMembers))
+            {
+                mergedBeforeRelation.withMembers(RelationBean.mergeBeans(leftMembers, rightMembers),
+                        Rectangle.forLocated(beforeRelationLeft, beforeRelationRight));
+            }
+            if (!leftOsmMembers.equalsIncludingExplicitlyExcluded(rightOsmMembers))
+            {
+                mergedBeforeRelation.withAllKnownOsmMembers(
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java]: Do we actually need to merge `allKnownOsmMembers`? I'm not sure if this field is ever inconsistent across shards. It sounds like it isn't.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java [hunk 2]
```diff
@@ -654,35 +605,55 @@ private static FeatureChange mergeRelations(final FeatureChange left, final Feat
...
+        final MergedMemberBean<RelationBean> mergedAllKnownMembersBean = new MemberMerger.Builder<RelationBean>()
+                .withMemberName("allKnownOsmMembers").withBeforeEntityLeft(beforeEntityLeft)
+                .withAfterEntityLeft(afterEntityLeft).withBeforeEntityRight(beforeEntityRight)
+                .withAfterEntityRight(afterEntityRight)
+                .withMemberExtractor(entity -> ((Relation) entity).allKnownOsmMembers() == null
+                        ? null : ((Relation) entity).allKnownOsmMembers().asBean())
+                .withAfterViewNoBeforeMerger(MemberMergeStrategies.simpleRelationBeanMerger)
+                .withAfterViewConsistentBeforeMerger(
+                        MemberMergeStrategies.diffBasedRelationBeanMerger)
+                .withAfterViewConflictingBeforeMerger(
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java]: Again, do we need to resolve conflicting beforeViews here? If `allKnownOsmMembers` is consistent across shards, then I think not.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java (hunk 1 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java]: It looks like it is not.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java (hunk 2 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChangeMergingHelpers.java]: Same I think...

File: src/main/java/org/openstreetmap/atlas/geography/atlas/change/MemberMerger.java [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+        try
+        {
+            afterMemberResult = this.afterViewConflictingBeforeMerger.apply(beforeMemberLeft,
+                    afterMemberLeft, beforeMemberRight, afterMemberRight);
+        }
+        catch (final Exception exception)
+        {
+            throw new CoreException(
+                    "Attempted afterViewConflictingBeforeMerge merge failed for {} with beforeView: {} vs {}; afterView: {} vs {}",
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/change/MemberMerger.java]: Is there a way to make this message more descriptive? It seems afterViewConflictingBeforeMerge is very implementation-specific, and when a user sees that in a stack trace, they wont understand

Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/change/MemberMerger.java]: I updated the message, let me know what you think. We can always change these messages in the future should we think of something better.

**Pull request merged**